jsign-crypto/src/main/java/net/jsign/jca/HashiCorpVaultSigningService.java

```diff
@@ -131,15 +198,81 @@ public byte[] sign(SigningServicePrivateKey privateKey, String algorithm, byte[]
+                shaType = "256";
+            }
+
+            request.put("input", Base64.getEncoder().encodeToString(data));
+            request.put("prehashed", true);
+            request.put("hash_algorithm", String.format("sha2-%s", shaType));
+
+            // By default, RSA sign in HashiCorp Vault Transit uses RSA-PSS.
+            if (privateKey.getAlgorithm().equals("RSA")) {
+                request.put("signature_algorithm", "pkcs1v15");
+            }
+        }
 
         try {
             Map<String, ?> response = client.post("sign/" + keyName, JsonWriter.format(request));
-            String signature = ((Map<String, String>) response.get("data")).get("signature");
+
+            if (response == null || !response.containsKey("data")) {
+                throw new IllegalArgumentException("Response or 'data' key is null.");
+            }
+
+            if (response.containsKey("warnings") && response.get("warnings") != null) {
+                logger.log(Level.WARNING, String.format("Vault responded with warnings: %s", response.get("warnings")));
+            }
+
+            if (response.containsKey("error")) {
+                throw new IllegalArgumentException(String.format("Vault responded with error: %s", response.get("error")));
+            }
+
+            @SuppressWarnings("unchecked") Map<String, String> response_data = (Map<String, String>) response.get("data");
+            if (response_data == null || !response_data.containsKey("signature")) {
+                throw new IllegalArgumentException("'data' or 'signature' key is null.");
+            }
+
+            String vault_signature = response_data.get("signature");
+            if (vault_signature == null) {
+                throw new IllegalArgumentException("Vault signature is null.");
+            }
+
+            /*
+             *  HashiCorp Vault Google Cloud KMS signature format   : '<BASE64_SIGNATURE>'
+             *  HashiCorp Vault Transit signature format            : 'vault:v<KEY_VERSION>:<BASE64_SIGNATURE>'
+             */
+            String signature;
+            if (this.googleCloudKMS) {
+                signature = vault_signature;
+            }
+            else {
+                String[] signatureParts = vault_signature.split(":");
+                if (signatureParts.length != 3) {
+                    throw new IllegalArgumentException("Invalid vault signature format.");
+                }
+                signature = signatureParts[2];
+            }
 
             return Base64.getDecoder().decode(signature);
+
+        } catch (IllegalArgumentException | ClassCastException e) {
+            logger.log(Level.SEVERE, "Invalid response format: ", e.getMessage());
+            throw new GeneralSecurityException("Invalid response format.", e);
         } catch (IOException e) {
-            throw new GeneralSecurityException(e);
+            logger.log(Level.SEVERE, "IO Exception during signing: ", e.getMessage());
```

Comment:
No need to log the exception, if it's thrown it will be logged or displayed elsewhere

Reply:
Removed the logging on this scenario in the next commit.